### PR TITLE
[SHELL32] Implement Copy To Folder

### PR DESCRIPTION
**Pull request by katahiromz**
## Purpose

JIRA issue: [CORE-11132](https://jira.reactos.org/browse/CORE-11132)

### CODE REVIEW

File: dll/win32/shell32/CCopyToMenu.cpp [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+
+    hr = SHGetPathFromIDListW(m_pidlFolder, wszPath);
+    if (FAILED_UNEXPECTEDLY(hr))
+        return hr;
+
+    BROWSEINFOW info = { lpici->hwnd };
+    info.pidlRoot = NULL;
+    info.lpszTitle = L"FIXME: This is a title";
+    info.ulFlags = BIF_RETURNONLYFSDIRS;
+    LPITEMIDLIST pidl = SHBrowseForFolder(&info);
```
Comment (learn-more) [inline dll/win32/shell32/CCopyToMenu.cpp:32]: Why not use a `CComHeapPtr<ITEMIDLIST>` ?

File: dll/win32/shell32/CCopyToMenu.cpp [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    m_pidlFolder(NULL),
+    m_idCmdFirst(0),
+    m_idCmdLast(0),
+    m_idCmdCopyTo(-1)
+{
+}
+
+CCopyToMenu::~CCopyToMenu()
+{
+    if (m_pidlFolder)
```
Comment (learn-more) [inline dll/win32/shell32/CCopyToMenu.cpp:15]: Why not using a `CComHeapPtr<ITEMIDLIST>` here?

File: dll/win32/shell32/CCopyToMenu.cpp (hunk 1 above)
Comment (katahiromz) [reply dll/win32/shell32/CCopyToMenu.cpp]: OK.

File: dll/win32/shell32/CCopyToMenu.cpp (hunk 2 above)
Comment (katahiromz) [reply dll/win32/shell32/CCopyToMenu.cpp]: OK.

**Comment by katahiromz** [comment]
There are both `CDefaultContextMenu::DoCopyToFolder` and `CCopyToMenu`. How can we connect them?

File: dll/win32/shell32/CCopyToMenu.cpp [hunk 3]
```diff
@@ -0,0 +1,173 @@
...
+    BROWSEINFOW info = { lpici->hwnd };
+    info.pidlRoot = NULL;
+    info.lpszTitle = strTitle;
+    info.ulFlags = BIF_RETURNONLYFSDIRS | BIF_USENEWUI;
+    info.lpfn = BrowseCallbackProc;
+    info.lParam = (LPARAM)this;
+    CComHeapPtr<ITEMIDLIST> pidl(SHBrowseForFolder(&info));
+    if (pidl)
+    {
+        // FIXME: Copy m_pDataObject into pidl
```
Comment (katahiromz) [inline dll/win32/shell32/CCopyToMenu.cpp:72]: @learn-more How to copy `m_pDataObject`?

Comment (learn-more) [reply dll/win32/shell32/CCopyToMenu.cpp:72]: Cannot copy from data object to pidl 1:1
A data object can contain multiple files, afaik a single pidl not.

Comment (learn-more) [reply dll/win32/shell32/CCopyToMenu.cpp:72]: Or do you want to create a data object from a pidl?

Comment (katahiromz) [reply dll/win32/shell32/CCopyToMenu.cpp:72]: Then how to drop the data object onto the pidl?

Comment (learn-more) [reply dll/win32/shell32/CCopyToMenu.cpp:72]: You probably want something like `CFontExt::Drop` is doing:
https://git.reactos.org/?p=reactos.git;a=blob;f=dll/shellext/fontext/CFontExt.cpp;hb=03e6eca3c2aaa61059085570f8e14039f401158f#l475

Comment (learn-more) [reply dll/win32/shell32/CCopyToMenu.cpp:72]: or if you mean apply the copy action for all files in the data object to pidl, then maybe `SHFileOperation`,
where you construct the `pFrom` with the data object items, and the `pTo` from the pidl?

File: dll/win32/shell32/CCopyToMenu.cpp [hunk 4]
```diff
@@ -0,0 +1,258 @@
...
+            SetWindowLongPtr(hwnd, GWLP_USERDATA, lpData);
+            this_ = reinterpret_cast<CCopyToMenu *>(lpData);
+
+            // Select initial directory
+            SendMessageW(hwnd, BFFM_SETSELECTION, FALSE,
+                reinterpret_cast<LPARAM>(static_cast<LPCITEMIDLIST>(this_->m_pidlFolder)));
+            break;
+
+        case BFFM_SELCHANGED:
+            pidl = reinterpret_cast<LPCITEMIDLIST>(lParam);
```
Comment (learn-more) [inline dll/win32/shell32/CCopyToMenu.cpp:70]: Who cleans up pidl?
Do we get a unique copy, or is it cleaned up by the one creating the `BFFM_SELCHANGED`

File: dll/win32/shell32/CCopyToMenu.cpp [hunk 5]
```diff
@@ -0,0 +1,258 @@
...
+            continue;
+
+        PIDLIST_ABSOLUTE pidlCombine = ILCombine(pidlParent, pidlRelative);
+        if (!pidl)
+            return E_FAIL;
+
+        SHGetPathFromIDListW(pidlCombine, szPath);
+        ILFree(pidlCombine);
+
+        if (PathFileExistsW(szPath))
```
Comment (learn-more) [inline dll/win32/shell32/CCopyToMenu.cpp:112]: Technically it does not have to be a real path, but we have this problem everywhere in our shell...

File: dll/win32/shell32/CDefaultContextMenu.cpp [hunk 6]
```diff
@@ -952,6 +954,29 @@ CDefaultContextMenu::DoCreateNewFolder(
...
+        return hr;
+
+    CComPtr<IContextMenu> pContextMenu;
+    hr = SHCoCreateInstance(NULL, &CLSID_CopyToMenu, NULL, IID_PPV_ARG(IContextMenu, &pContextMenu));
+    if (FAILED(hr))
+        return hr;
+
+    CComPtr<IShellExtInit> pInit;
+    hr = pContextMenu->QueryInterface(IID_PPV_ARG(IShellExtInit, &pInit));
+    if (FAILED(hr))
```
Comment (learn-more) [inline dll/win32/shell32/CDefaultContextMenu.cpp:970]: `FAILED_UNEXPECTEDLY`?

File: dll/win32/shell32/CCopyToMenu.cpp (hunk 4 above)
Comment (katahiromz) [reply dll/win32/shell32/CCopyToMenu.cpp:70]: The procedure is not owner of this pidl. We don't have to clean it up.

File: dll/win32/shell32/CDefaultContextMenu.cpp (hunk 6 above)
Comment (katahiromz) [reply dll/win32/shell32/CDefaultContextMenu.cpp]: OK.

File: dll/win32/shell32/CCopyToMenu.cpp (hunk 5 above)
Comment (katahiromz) [reply dll/win32/shell32/CCopyToMenu.cpp]: OK.

**Comment by katahiromz** [comment]
Buggy and not working.
hungup
hungup2
```txt
FramePtr ReturnAd Param#1  Param#2  Param#3  Param#4  Function Name
0198F5C4 112DF40C 00000000 00000000 00000000 00000000 <nomod>!<nosymbols> +0x0
0198F5DC 7B0E09D8 0028DA10 0028DA10 0198F61C 7B07EC52 shell32!_ZN3ATL10CComObjectI19CDefaultContextMenuED0Ev +0x28
0198F5EC 7B0E09A4 0025C9C0 00000000 00000000 00000000 shell32!_ZN3ATL10CComObjectI19CDefaultContextMenuE7ReleaseEv +0x1c
0198F61C 7B07EC52 00000000 00000000 00000000 00000000 shell32!CDefView::OpenSelectedItems +0x141
0198F794 7B081799 0000004E 00000001 0198FB34 0198F7B0 shell32!CDefView::OnNotify +0x384
0198F7C4 7B0E81BD 0018013A 0000004E 00000001 0198FB34 shell32!CDefView::ProcessWindowMessage +0x298
```

**Comment by katahiromz** [comment]
```txt
err:(dll/win32/shell32/CDefView.cpp:2258) no menu!!!
```
What's?

**Comment by katahiromz** [comment]
contextmenu
ContextMenuHandler is not working...

**Comment by katahiromz** [comment]
```txt
err:(dll/win32/shell32/CCopyToMenu.cpp:173) CCopyToMenu::QueryContextMenu(000A0084, 2, 3, 32767, 0)
err:(win32ss/user/user32/windows/menu.c:608) invalid combination of fMask bits used
err:(dll/win32/shell32/CCopyToMenu.cpp:173) CCopyToMenu::QueryContextMenu(000B0084, 2, 3, 32767, 0)
err:(win32ss/user/user32/windows/menu.c:608) invalid combination of fMask bits used
```

**Comment by katahiromz** [comment]
contextmenu
Display of menu item is OK. Action is not working...
```txt
err:(dll/win32/shell32/CCopyToMenu.cpp:173) CCopyToMenu::QueryContextMenu(00090082, 2, 3, 32767, 0)
err:(dll/win32/shell32/CCopyToMenu.cpp:173) CCopyToMenu::QueryContextMenu(000A0082, 2, 3, 32767, 0)
err:(dll/win32/shell32/CDefaultContextMenu.cpp:1281) Unhandled Verb 7ffel
CDefView.cpp:1413: Unexpected failure 80070057.
```

**Comment by katahiromz** [comment]
itworks
The context menu is working...!

**Comment by katahiromz** [comment]
editnotworking
"Edit" menu not working...

**Comment by katahiromz** [comment]
fully-working
Now fully working. Context menu is OK. The "Edit" menu is OK. "OK" button is OK. Copying is OK. All rights!

File: dll/win32/shell32/lang/es-ES.rc [hunk 7]
```diff
@@ -992,4 +992,6 @@ BEGIN
     IDS_EDITING_ACTION "Editar acción: "
     IDS_NO_ICONS "El archivo '%s' no contiene íconos.\n\nEscoja un ícono de la lista o seleccione otro archivo."
     IDS_FILE_NOT_FOUND "El archivo '%s' no pudo ser encontrado."
+    IDS_COPYTOMENU "Copy To &Folder..."
```
[the remark this answers is not in the document]
Comment (binarymaster) [reply dll/win32/shell32/lang/es-ES.rc:995]: @julenuri we usually add translations later

**Pull request closed without merging**